### PR TITLE
Add FanartTV and caching

### PR DESCRIPTION
**Pull request by AMoo-Miki**
Code is shamelessly borrowed from metadata.tvshows.themoviedb.org.python

### CODE REVIEW

File: resources/lib/cache.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
+# -*- coding: UTF-8 -*-
```
Comment (romanvm) [inline resources/lib/cache.py:1]: No need to use `coding` declaration in Python 3.

File: resources/lib/cache.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    file_name = str(show_id) + '-' + lang + '.pickle'
+    cache = load_show_info_from_cache(show_id, lang)
+    if cache:
+        cache[part] = data
+    else:
+        cache = {
+            part: data
+        }
+    with open(os.path.join(CACHE_DIR, file_name), 'wb') as fo:
+        pickle.dump(cache, fo, protocol=2)
```
Comment (romanvm) [inline resources/lib/cache.py:65]: Pickle-based cache proved to be unreliable across Python versions and even across Python 3 minor versions. This is why I had to switch to JSON-based cache in my TVmaze scraper: https://github.com/romanvm/kodi.tvmaze/blob/master/metadata.tvmaze/libs/cache_service.py
In Python 3 you won't need to do explicit encoding before saving JSON and can just use `json.dump()`.

File: resources/lib/cache.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    if not xbmcvfs.exists(cache_dir):
+        xbmcvfs.mkdir(cache_dir)
+    logger.debug('the cache dir is ' + cache_dir)
+    return cache_dir
+
+
+CACHE_DIR = _get_cache_directory()  # type: Text
+
+
+def cache_show_info(show_id, data, lang, part):
```
Comment (romanvm) [inline resources/lib/cache.py:48]: I don't see that you set/check cache TTL anywhere. Do you want to cache data forever?

File: addon.xml [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
-<addon id="metadata.tvdb.com.python" name="The TVDB (new)" version="1.0.3" provider-name="Team Kodi">
+<addon id="metadata.tvdb.com.python" name="The TVDB (new)" version="1.0.4" provider-name="Team Kodi">
   <requires>
     <import addon="xbmc.metadata" version="2.1.0" />
     <import addon="xbmc.python" version="3.0.0" />
     <import addon="script.module.tvdbsimple" version="1.0.6" />
     <import addon="script.module.trakt" version="3.1.0" />
   </requires>
-  <extension point="xbmc.metadata.scraper.tvshows" language="en" library="scraper.py" />
+  <extension point="xbmc.metadata.scraper.tvshows" language="en" library="scraper.py" cachepersistence="00:15" />
```
Comment (romanvm) [inline addon.xml:9]: I have doubts that this option has any effect on Python scrapers.

File: resources/lib/fanarttv.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python
```
Comment (romanvm) [inline resources/lib/fanarttv.py:1]: Shebang is meant for executable scripts in *nix environment, not for importable modules.

File: resources/lib/fanarttv.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+# -*- coding: UTF-8 -*-
```
Comment (romanvm) [inline resources/lib/fanarttv.py:2]: Again, no need for coding declaration in Python 3.

File: resources/lib/fanarttv.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            return
+
+        resp = json.loads(response.read().decode('utf-8'))
+        cache.cache_fanarttv_info(id, resp)
+
+    language = settings.getSettingString('language')
+
+    for key in resp:
+        subKey = 'artwork'
+        if key == 'hdtvlogo' or key == 'clearlogo':
```
Comment (romanvm) [inline resources/lib/fanarttv.py:58]: It is better to use a dictionary mapping here instead of a cascade of elifs.

**Review by romanvm — CHANGES_REQUESTED** [review]
I strongly recommend to rework caching. Other changes are optional.

File: resources/lib/cache.py (hunk 2 above)
Comment (AMoo-Miki) [reply resources/lib/cache.py:65]: Thanks for looking at this. No part of the caching or fanarttv part is mine; all is stolen. I will go steal yours and update this :D

File: addon.xml (hunk 4 above)
Comment (Karlson2k) [reply addon.xml:9]: It works for any kind of scraper.

File: resources/lib/cache.py (hunk 3 above)
Comment (Karlson2k) [reply resources/lib/cache.py:48]: If you put files to the right location then kodi checks for any outdated files (and deletes them) each time before plugin started.

Comment (romanvm) [reply resources/lib/cache.py:48]: Ok, I didn't know about that feature. Then I withdraw my comments about cache TTL setting. The caching mechanism itself still needs to be reworked, pickling is too unreliable.

**Pull request closed without merging**